File: backend/src/main/java/br/com/calciolari/datahub/imports/infrastructure/persistence/ParsedMovementEntity.java
package br.com.calciolari.datahub.imports.infrastructure.persistence;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.UUID;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.Id;
import jakarta.persistence.Table;

@Entity
@Table(name = "parsed_movement")
public class ParsedMovementEntity {

	@Id
	private UUID id;

	@Column(name = "parse_attempt_id", nullable = false)
	private UUID parseAttemptId;

	@Column(name = "source_record_index", nullable = false)
	private int sourceRecordIndex;

	@Column(nullable = false)
	private String direction;

	@Column(name = "external_product_id")
	private String externalProductId;

	@Column(name = "product_name")
	private String productName;

	@Column(name = "external_sale_id")
	private String externalSaleId;

	@Column(name = "occurred_at")
	private LocalDateTime occurredAt;

	private BigDecimal quantity;

	@Column(name = "unit_price")
	private BigDecimal unitPrice;

	@Column(name = "discount_percentage")
	private BigDecimal discountPercentage;

	private BigDecimal total;

	@Column(name = "previous_stock")
	private BigDecimal previousStock;

	@Column(name = "resulting_stock")
	private BigDecimal resultingStock;

	private String manufacturer;

	@Column(name = "source_locator")
	private String sourceLocator;

	protected ParsedMovementEntity() {
	}

	public static ParsedMovementEntity from(
			UUID id,
			UUID parseAttemptId,
			br.com.calciolari.datahub.imports.domain.parser.ParsedMovement movement) {
		ParsedMovementEntity entity = new ParsedMovementEntity();
		entity.id = id;
		entity.parseAttemptId = parseAttemptId;
		entity.sourceRecordIndex = movement.sourceRecordIndex();
		entity.direction = movement.direction().name();
		entity.externalProductId = movement.externalProductId();
		entity.productName = movement.productName();
		entity.externalSaleId = movement.externalSaleId();
		entity.occurredAt = movement.occurredAt();
		entity.quantity = movement.quantity();
		entity.unitPrice = movement.unitPrice();
		entity.discountPercentage = movement.discountPercentage();
		entity.total = movement.total();
		entity.previousStock = movement.previousStock();
		entity.resultingStock = movement.resultingStock();
		entity.manufacturer = movement.manufacturer();
		if (movement.sourceLocator() != null) {
			entity.sourceLocator = String.valueOf(movement.sourceLocator());
		}
		return entity;
	}

	public UUID getId() {
		return id;
	}

	public UUID getParseAttemptId() {
		return parseAttemptId;
	}

	public int getSourceRecordIndex() {
		return sourceRecordIndex;
	}

	public String getDirection() {
		return direction;
	}

	public String getExternalProductId() {
		return externalProductId;
	}

	public String getProductName() {
		return productName;
	}

	public String getExternalSaleId() {
		return externalSaleId;
	}

	public LocalDateTime getOccurredAt() {
		return occurredAt;
	}

	public BigDecimal getQuantity() {
		return quantity;
	}

	public BigDecimal getUnitPrice() {
		return unitPrice;
	}

	public BigDecimal getDiscountPercentage() {
		return discountPercentage;
	}

	public BigDecimal getTotal() {
		return total;
	}
}

File: backend/src/Calciolari.DataHub/Persistence/Entities/ParsedMovementEntity.cs
using Calciolari.DataHub.Imports.Domain.Parser;

namespace Calciolari.DataHub.Persistence.Entities;

public sealed class ParsedMovementEntity
{
    public Guid Id { get; set; }
    public Guid ParseAttemptId { get; set; }
    public int SourceRecordIndex { get; set; }
    public string Direction { get; set; } = string.Empty;
    public string? ExternalProductId { get; set; }
    public string? ProductName { get; set; }
    public string? ExternalSaleId { get; set; }
    public DateTime? OccurredAt { get; set; }
    public decimal? Quantity { get; set; }
    public decimal? UnitPrice { get; set; }
    public decimal? DiscountPercentage { get; set; }
    public decimal? Total { get; set; }
    public decimal? PreviousStock { get; set; }
    public decimal? ResultingStock { get; set; }
    public string? Manufacturer { get; set; }
    public string? SourceLocator { get; set; }

    public static ParsedMovementEntity From(Guid id, Guid parseAttemptId, ParsedMovement movement) =>
        new()
        {
            Id = id,
            ParseAttemptId = parseAttemptId,
            SourceRecordIndex = movement.SourceRecordIndex,
            Direction = movement.Direction.WireName(),
            ExternalProductId = movement.ExternalProductId,
            ProductName = movement.ProductName,
            ExternalSaleId = movement.ExternalSaleId,
            OccurredAt = movement.OccurredAt,
            Quantity = movement.Quantity,
            UnitPrice = movement.UnitPrice,
            DiscountPercentage = movement.DiscountPercentage,
            Total = movement.Total,
            PreviousStock = movement.PreviousStock,
            ResultingStock = movement.ResultingStock,
            Manufacturer = movement.Manufacturer,
            SourceLocator = movement.SourceLocator.ToString()
        };
}
